Ternary expressions with null fail

The following program is no longer considered valid under v2.1.0:

```java
import javax.annotation.Nullable;

public class Checker {
  @Nullable
  public static Double getDouble(boolean flag) {
    return flag ? null : 1.0;
  }
}
```

This works under v2.0.0. The error message is:

```
Checker.java:8: error: [type.invalid] [@Initialized, @Nullable] may not be applied to the type "@Initialized @Nullable double"
    return flag ? null : 1.0;
```

--- a/checker/tests/nullness/Issue829.java
+++ b/checker/tests/nullness/Issue829.java
@@ -1,6 +1,5 @@
 // Test case for Issue 829
 // https://github.com/typetools/checker-framework/issues/829
-// @skip-test
 
 import org.checkerframework.checker.nullness.qual.*;
 

--- a/framework/src/org/checkerframework/framework/type/TypeFromExpressionVisitor.java
+++ b/framework/src/org/checkerframework/framework/type/TypeFromExpressionVisitor.java
@@ -107,19 +107,26 @@ class TypeFromExpressionVisitor extends TypeFromTreeVisitor {
     public AnnotatedTypeMirror visitConditionalExpression(
             ConditionalExpressionTree node, AnnotatedTypeFactory f) {
 
+        // The Java type of a conditional expression is generally the LUB of the boxed types
+        // of the true and false expressions, but with a few exceptions. See JLS 15.25.
+        // So, use the type of the ConditionalExpressionTree instead of InternalUtils#leastUpperBound
+        AnnotatedTypeMirror alub = f.type(node);
+
         AnnotatedTypeMirror trueType = f.getAnnotatedType(node.getTrueExpression());
         AnnotatedTypeMirror falseType = f.getAnnotatedType(node.getFalseExpression());
+
         if (trueType.getKind() == TypeKind.NULL) {
-            return AnnotatedTypes.lubWithNull((AnnotatedNullType) trueType, falseType, f);
+            AnnotatedTypeMirror falseTypeAsLub = AnnotatedTypes.asSuper(f, falseType, alub);
+            return AnnotatedTypes.lubWithNull((AnnotatedNullType) trueType, falseTypeAsLub, f);
         } else if (falseType.getKind() == TypeKind.NULL) {
-            return AnnotatedTypes.lubWithNull((AnnotatedNullType) falseType, trueType, f);
+            AnnotatedTypeMirror trueTypeAsLub = AnnotatedTypes.asSuper(f, trueType, alub);
+            return AnnotatedTypes.lubWithNull((AnnotatedNullType) falseType, trueTypeAsLub, f);
         }
 
         if (trueType.equals(falseType)) {
             return trueType;
         }
 
-        AnnotatedTypeMirror alub = f.type(node);
 
         trueType = AnnotatedTypes.asSuper(f, trueType, alub);
         falseType = AnnotatedTypes.asSuper(f, falseType, alub);

